src/program/snabbvmx/query/query.lua

```diff
@@ -0,0 +1,154 @@
+   local counters = shm.children(counters_path)
+   for _, name in ipairs(counters) do
+      cnt_path = counters_path .. name
+      cnt = counter.open(cnt_path, 'readonly')
+      value = tonumber(counter.read(cnt))
+      name = name:gsub(".counter$", "")
+      ret[name] = value
+    end
+   return ret
+end
+
+local function read_apps_counters (tree, app_name)
+   local ret = {}
+   local cnt, cnt_path, value
+   local counters_path = "/" .. tree .. "/apps/" .. app_name .. "/"
+   local counters = shm.children(counters_path)
+   for _, name in ipairs(counters) do
+      cnt_path = counters_path .. name
+      cnt = counter.open(cnt_path, 'readonly')
+      value = tonumber(counter.read(cnt))
+      name = name:gsub(".counter$", "")
+      ret[name] = value
+    end
+   return ret
+end
+
+local function print_counter (name, value)
+   print(("      <%s>%d</%s>"):format(name, value, name))
+end
+
+function print_lwaftr_counters (tree)
+   -- Open, read and print whatever counters are in that directory.
+   local counters = read_lwaftr_counters(tree)
+   for _, name in ipairs(sort(keys(counters))) do
+     local value = counters[name]
+     print_counter(name, value)
+   end
+end
+
+-- TODO: Refactor to a general common purpose library.
+local function file_exists(path)
+  local stat = S.stat(path)
+  return stat and stat.isreg
+end
+
+function print_next_hop (pid, name)
+  local next_hop_mac = "/" .. pid .. "/" .. name
+  if file_exists(shm.root .. next_hop_mac) then
+    local nh = shm.open(next_hop_mac, macaddress_t, "readonly")
+    print(("    <%s>%s</%s>"):format(name, ethernet:ntop(nh.ether), name))
+  end
+end
+
+function print_monitor (pid)
+  local path = "/" .. pid .. "/v4v6_mirror"
+  if file_exists(shm.root .. path) then
+    local ipv4_address = shm.open(path, "struct { uint32_t ipv4; }", "readonly")
+    print(("    <%s>%s</%s>"):format("monitor", ipv4:ntop(ipv4_address), "monitor"))
+  end
+end
+
+function print_apps_counters (tree)
+  local apps_path = "/" .. tree .. "/apps"
+   local apps = shm.children(apps_path)
+   for _, app_name in ipairs(apps) do
+     print(("    <%s>"):format(app_name))
+     -- Open, read and print whatever counters are in that directory.
+     local counters = read_apps_counters(tree, app_name)
+     for _, name in ipairs(sort(keys(counters))) do
+       local value = counters[name]
+       print_counter(name, value)
+     end
+     print(("    </%s>"):format(app_name))
+   end
+end
+
+function run (raw_args)
+   parse_args(raw_args)
+   print("<snabb>")
+   for _, pid in ipairs(shm.children("/")) do
+      if shm.exists("/"..pid.."/nic/id") then
+         local lwaftr_id = shm.open("/"..pid.."/nic/id", lwtypes.lwaftr_id_type)
+         local instance_id = ffi.string(lwaftr_id.value)
+         if instance_id then
+           print("  <instance>")
+           print(("   <instance>%s</instance>"):format(instance_id))
+           print(("   <name>%s</name>"):format(ffi.string(lwaftr_id.value)))
```

Comment:
`instance_id` and `ffi.string(lwaftr_id.value)` are the same value, aren't they? What should be the difference between `instance` and `name` tags?